Implement multiple selection & hidden folders for wxDirDialog

imciner2 opened the pull request:
This PR adds GTK and OSX support for both multiple selection and hidden folders into wxDirDialog. The behavior used here is consistent with `wxFileDialog`. Additionally, the OSX wxDirDialog implementation has been rewritten to use the non-deprecated API for NSOpenPanel, and fix an issue with the title of the dialog not being set.

Fixes [#18736](https://trac.wxwidgets.org/ticket/18736), [#15143](https://trac.wxwidgets.org/ticket/15143)

cc @PBfordev

imciner2 committed:
Add wxDD_MULTIPLE wxDirDialog style and implement it for GTK

Add wxDirDialog::GetPaths() similar to the existing member of
wxFileDialog with the same name and also taking, for consistency,
wxArrayString as the output parameter.

imciner2 committed:
Update sample to use multiple selection in wxDirDialog

imciner2 committed:
Add show hidden folders flag to wxDirDialog

Add wxDD_SHOW_HIDDEN similar to the existing wxFD_SHOW_HIDDEN.

imciner2 committed:
Implement hidden directory support in wxOSX too

Add support for wxDD_SHOW_HIDDEN style to wxDirDialog in wxOSX.

imciner2 committed:
Implement multiple selection support in wxOSX

Add support for wxDD_MULTIPLE style to wxDirDialog in wxOSX too.

imciner2 committed:
Use non-deprecated NSOpenPanel methods in wxDirDialog on OSX

imciner2 committed:
Fix setting the title for the wxDirDialog on OSX

OSX 10.11+ doesn't actually display the title, so update documentation
to reference SetMessage instead. For pre-10.11 override the SetTitle
method to set the title of the NSOpenPanel instead of the window.

Also change so the directory is not updated unless it is provided.

Closes #15143.

imciner2 committed:
Warn on incompatible wxDirDialog styles

PBfordev reviewed (commented):
I added few comments and suggestions to the documentation and even GTK and MacOS specific parts. Hope you do not mind.

PBfordev commented:
I believe that new `wxDD_` flags should be implemented in a single PR for all platforms, so it can be easily tested with CI and merged at once.

I have pushed the MSW implementation against your branch. If you have a better idea how to merge the MSW implementation with your PR, please let me know.

imciner2 commented on src/gtk/dirdlg.cpp:
```diff
@@ -71,6 +71,8 @@ bool wxDirDialog::Create(wxWindow* parent,
 {
     m_message = title;
 
+    wxASSERT( !( HasFlag(wxDD_MULTIPLE) && HasFlag(wxDD_CHANGE_DIR) ) );
```
Good point, the window flags are set by the `wxWindowBase::CreateBase` function (https://github.com/wxWidgets/wxWidgets/blob/master/src/common/wincmn.cpp#L392), which is called after this point.

imciner2 commented on interface/wx/dirdlg.h:
```diff
@@ -43,6 +48,13 @@ const char wxDirDialogNameStr[] = "wxDirCtrl";
+           @note This flag is ignored if the @c wxDD_MULTIPLE style is also used
+    @style{wxDD_MULTIPLE}
+           Allow the user to select multiple directories.
+           @since 3.1.4
```
I have seen this kind of formatting issue before, but I think they aren't present on the doxygen version that is used to generate the actual docs (since they don't appear in the online docs). Perhaps @vadz can shed some light on this.

imciner2 replied:
The ones on the wxGauge documentation page you linked look correct though, since they don't cause issue with the future spacing. I would guess this is an issue that was created with later Doxygen versions, so I will go ahead and make it a sentence instead of a macro for better compatibility.

imciner2 replied:
I have removed the `@since` usage and made them a separate sentence.

imciner2 commented:
> I believe that new `wxDD_` flags should be implemented in a single PR for all platforms, so it can be easily tested with CI and merged at once.
> 
> I have pushed the MSW implementation against your branch. If you have a better idea how to merge the MSW implementation with your PR, please let me know.

I don't think you can push directly to my branch yet. I have invited you to my fork of this repo, so once you accept that you should be able to push your commits onto this branch (`im/dirdialog`) in my repository and have them be in this same pull request.

PBfordev committed:
Implement MSW support for wxDD_MULTIPLE and wxDD_SHOW_HIDDEN

PBfordev commented:
Thanks for the invite. I am GIT-impaired so I just created a PR against your branch and then merged it via GitHub. Hopefully everything went well.

vadz commented on src/gtk/dirdlg.cpp:
```diff
@@ -132,13 +143,24 @@ bool wxDirDialog::Create(wxWindow* parent,
+    if ( !HasFlag(wxDD_MULTIPLE) && HasFlag(wxDD_CHANGE_DIR) )
     {
-        wxSetWorkingDirectory(m_selectedDirectory);
+        wxSetWorkingDirectory(m_paths.Last());
```
Just checking: are we really guaranteed to have a path here? I.e. it's impossible to accept the dialog without choosing anything?

imciner2 replied:
When I test this on the dialogs sample on GTK3, it is giving the directory that the dialog is in as the returned directory if the user didn't select anything.

vadz commented on src/msw/dirdlg.cpp:
```diff
@@ -212,38 +146,54 @@ void wxDirDialog::SetPath(const wxString& path)
 
     // change current working directory if asked so
-    if ( rc == wxID_OK && HasFlag(wxDD_CHANGE_DIR) )
+    // but only if the dialog is single selection only
+    if ( rc == wxID_OK &&
+         ( HasFlag(wxDD_CHANGE_DIR) && !HasFlag(wxDD_MULTIPLE) ) )
```
I'm not convinced we should do this. We've already given them an assert in ctor to say that this combination is not allowed. And it doesn't matter what we do when preconditions are not satisfied, so why complicate the code with more checks. Personally, I'd just leave the original code unchanged and rely on `wxDD_CHANGE_DIR` presence implying `wxDD_MULTIPLE` absence.

PBfordev replied:
OK, will revert.

I just look at things from the user perspective, where it can mean that the CWD is changed to a seemingly random folder. Surprisingly, some people even do not use the debug versions of libraries for development...

vadz replied:
They don't have to use debug build of wx for development and they don't have to read the docs, which document that this combination is unsupported, neither, but I don't think we have to go to any special trouble to support them neither...

vadz commented on src/msw/dirdlg.cpp:
```diff
@@ -309,39 +259,94 @@ int wxDirDialog::ShowSHBrowseForFolder(WXHWND owner)
+            wxLogApiError(wxS("IFileDialog::Show"), hr);
+        }
+    }
+    else
+    if ( GetPathsFromIFileOpenDialog(fileDialog, HasFlag(wxDD_MULTIPLE),
```
```suggestion
    else if ( GetPathsFromIFileOpenDialog(fileDialog, HasFlag(wxDD_MULTIPLE),
```
Just a style nitpick.

vadz commented on src/msw/dirdlg.cpp:
```diff
@@ -378,77 +383,122 @@ int wxDirDialog::ShowIFileDialog(WXHWND owner)
+    if ( tempPaths.empty() )
+        return false; // there was en error
+
+    paths = tempPaths;
```
Could do
```suggestion
    paths.swap(tempPaths);
```
for a tiny efficiency gain. Not really sure if it's worth it.

PBfordev replied:
I am aware of the trick, but I preferred to put emphasis on intent, i.e., one direction assignment and not a swap, believing the possible performance benefit negligible.

vadz commented on src/msw/dirdlg.cpp:
```diff
@@ -378,77 +383,122 @@ int wxDirDialog::ShowIFileDialog(WXHWND owner)
+    }
+
+    path = pathOLE;
+    CoTaskMemFree(pathOLE);
```
Not really meant for this PR, but we should have a smart pointer releasing memory using `::CoTaskMemFree()`.

PBfordev replied:
I grepped the src folder for `CoTaskMemFree()`.  It is used once by `wxDirDialog`, `wxStandardPaths`, and `wxTaskBarButton`; five times by `wxWebViewEdgeImpl`.

Not sure if this demands creating a new class but I could look into it if you think it would be useful.

vadz replied:
Well, ideally, yes, I think it would: the rule is to never deallocate any resources manually and always use RAII for them, and this definitely falls under it.

Whether it's worth it compared to the other possible improvements is a different question and, again, not doing it definitely won't prevent this PR from being merged, but it would be nice to not have "raw" calls to free memory.

PBfordev replied:
Well, I did add `wxCOMPtr` when adding support for Vista+ style `wxDirDialog` back in 2012, but in this case it can IMO wait till a separate PR. 

ATM, I am not even sure how to name such a smart pointer and what its API should look like.

vadz commented on src/msw/dirdlg.cpp:
```diff
@@ -309,39 +259,94 @@ int wxDirDialog::ShowSHBrowseForFolder(WXHWND owner)
+    }
+
+    // Failed to show the dialog or obtain the selected folders(s)
+    wxLogSysError(_("Couldn't obtain folder name"), hr);
```
The error message doesn't seem ideal, but I admit I have trouble finding better suggestions. It probably doesn't matter anyhow, but in a highly unlikely event this really happens, it would be nice to somehow link it with `wxDirDialog`, so maybe it should be "Couldn't obtain the result from the directory selection dialog" or something like that?

PBfordev replied:
We already talked about it. I have been reluctant to make just a small change when the string then needs to be re-translated in all the languages.

I will change the string to what you suggested, if you think it is for the better.

vadz replied:
Sorry, I somehow totally failed to see that this was an existing message. Yes, I agree that it's not worth changing the translations for a message that nobody is going to see anyhow. Sorry again for wasting your time.

vadz commented on src/osx/cocoa/dirdlg.mm:
```diff
@@ -62,26 +65,40 @@
 WX_NSOpenPanel wxDirDialog::OSXCreatePanel() const
 {
-    NSOpenPanel *oPanel = [NSOpenPanel openPanel];
+    NSOpenPanel* oPanel = [NSOpenPanel openPanel];
```
If possible, please avoid whitespace-only changes.

imciner2 replied:
In this case, I was making a lot of changes in the file and I had originally made a change here that I ended up not needing. The whitespace change was a byproduct of that. I don't think it needs feverting in the review since that will cause more noise.

vadz commented on src/gtk/dirdlg.cpp:
```diff
@@ -132,13 +143,24 @@ bool wxDirDialog::Create(wxWindow* parent,
 
     // change to the directory where the user went if asked
-    if (HasFlag(wxDD_CHANGE_DIR))
+    if ( !HasFlag(wxDD_MULTIPLE) && HasFlag(wxDD_CHANGE_DIR) )
```
I'd leave this unchanged for the same reasons as in wxMSW.

imciner2 replied:
Fixed.

vadz commented on src/osx/cocoa/dirdlg.mm:
```diff
@@ -62,26 +65,40 @@
+    wxCFStringRef cf(m_message);
     [oPanel setMessage:cf.AsNSString()];
 
+    if ( !m_title.IsEmpty() )
```
```suggestion
    if ( !m_title.empty() )
```
Very minor, but it's preferable to use standard-like methods in the new code.

imciner2 replied:
Fixed.

imciner2 replied:
(although I always get confused if these functions that don't begin with `Is` will empty the string or just test for empty).

vadz reviewed (commented):
Thanks a lot for your work!

Globally this looks perfectly fine to me, there are just some minor fixes/possible simplifications that we could apply before merging.

I also didn't repeat the comments from #1883, but they still apply, i.e. please just remove `wxT()` and make the assert messages more prescriptive here too.

TIA!

PBfordev commented:
I have made the changes in the MSW code requested by vadz, but I do not know how to add them to this PR. GitHub does not let me merge my branch with Ian's and when I try creating a PR as I did before, it contains all my previous commits I already pushed there.

I apologize for my incompetence, the commit in question can be found here: https://github.com/wxWidgets/wxWidgets/commit/6127011353774fd2c71b4ed18e9eda57650cf89c

vadz commented:
It would be nice to finish this before 3.1.4. Ian, please let me know if you plan to do it during the next week or so or if I should try to address the remaining issues myself. TIA!

imciner2 commented on include/wx/dirdlg.h:
```diff
@@ -25,8 +25,24 @@ extern WXDLLIMPEXP_DATA_CORE(const char) wxDirDialogNameStr[];
+    wxDD_DIR_MUST_EXIST = 0x0200,
+    wxDD_MULTIPLE       = 0x0400,
+    wxDD_SHOW_HIDDEN    = 0x0001
+};
```
I changed this back to `#define`.

imciner2 commented on include/wx/osx/dirdlg.h:
```diff
@@ -47,6 +47,14 @@ class WXDLLIMPEXP_CORE wxDirDialog : public wxDirDialogBase
     virtual int ShowModal() wxOVERRIDE;
 
+    // MacOS 10.11 has removed the titlebar from the dialog, so this is provided
+    // for compatibility with older versions
```
Fixed.

imciner2 commented:
I have gone through the OSX and GTK codebase (and the headers) to address the comments. Let me know if there is anything else that should be updated.

> I also didn't repeat the comments from #1883, but they still apply, i.e. please just remove wxT() and make the assert messages more prescriptive here too.

I have now addressed these here in the last two commits for OSX and GTK.

vadz commented:
Thanks, looks good to me. I'd prefer to use `clear()` instead of `Empty()` for consistency with the standard containers and to avoid confusion with `empty()`, but I can change this myself. Let's just wait for the CI builds to finish to see if there are any problems.

Thanks again to both of you!

vadz commented:
BTW, I plan rebasing the commits a little to tidy up history and avoid all the "fixes" commits. I hope you don't mind, but please let me know if you do (e.g. if you thought about something else you'd like to add to this branch).

tomay3000 commented on include/wx/gtk/dirdlg.h:
```diff
@@ -55,7 +56,7 @@ class WXDLLIMPEXP_CORE wxDirDialog : public wxDirDialogBase
 
 private:
-    wxString m_selectedDirectory;
+    wxArrayString m_paths;
```
is `wxArrayString` still alive! I thought it is better replaced with `wxVector<wxString>`.

imciner2 commented:
> BTW, I plan rebasing the commits a little to tidy up history and avoid all the "fixes" commits. I hope you don't mind, but please let me know if you do (e.g. if you thought about something else you'd like to add to this branch).

I don't think I have anything else to add to this, so go ahead and clean up as needed.

PBfordev commented:
> please let me know if you do (e.g. if you thought about something else you'd like to add to this branch).

Perhaps you could also consider replacing raw pointer in `ConvertIShellItemToPath()` in _src/msw/dirdlg.cpp_ with `CoTaskMemPtr`? 

Due to my inability to work with GIT I thought I would do that as a separate PR once this one was merged but I guess it would be better if all the changes could be merged at once...

vadz commented:
> Perhaps you could also consider replacing raw pointer in `ConvertIShellItemToPath()` in _src/msw/dirdlg.cpp_ with `CoTaskMemPtr`?

Yes, I thought about doing this too, thanks for the reminder.

PBfordev commented:
I have realized that the new style for showing hidden folders (and files) has not been made available to `wxDirPicker` (and `wxFilePicker`). Is this intentional? Granted, showing hidden folders/files is not that common...

vadz commented:
> I have realized that the new style for showing hidden folders (and since 3.1.3 files ) has not been made available to `wxDirPicker` (and `wxFilePicker`) . Is it planned in a separate PR, or just thought not important as showing hidden folders/files is not that common?

I think it was just an oversight, not a conscious decision not to support the new styles there, so, as usual, any changes doing it would be welcome (either here or in a new PR once I finally find time to merge this one). FWIW I almost certainly won't have time to do it myself before 3.1.4, but you probably already knew that.

PBfordev commented:
OK, I will try to do that.

imciner2 commented:
I don't have time to do anything with the picker controls, and they are not high on my priority list because the application I work on doesn't use them at all (we only use the dialogs). I would suggest anything with the pickers be done in a new PR and not hold this one up.

PBfordev commented:
I am a bit busy ATM, so I have not done any work on supporting the new styles in pickers but I intend to do that in a separate PR.

vadz commented on src/osx/cocoa/dirdlg.mm:
```diff
@@ -132,18 +145,42 @@
+    if ( returnCode == NSOKButton )
     {
         NSOpenPanel* oPanel = (NSOpenPanel*)panel;
-        SetPath( wxCFStringRef::AsStringWithNormalizationFormC([[oPanel filenames] objectAtIndex:0]));
```
Please don't change this now, as I'm going to push my rebased version soon, but I wonder about this: is it correct that we "lost" `AsStringWithNormalizationFormC()` call in the new version? Does `[NSURL fileSystemRepresentation]` take care of this?

vadz committed:
Use wxCoTaskMemPtr<> instead of manual ::CoTaskMemFree()

No real changes, just use a smart pointer instead of manual memory
management calls.

vadz committed:
Avoid using wxString::Empty()

This is confusingly similar to std::string::empty() which doesn't do the
same thing, so prefer using clear() instead.

And simply remove Empty() calls before the assignment, as they're
useless.

No real changes.

vadz committed:
Avoid overriding wxDirDialog::GetPath[s]() unnecessarily

Don't duplicate practically the same code in all ports, just add m_paths
itself to the base class. The only drawback of doing this is that it's
unused in the ports not (yet) using it, but this saves enough code in
the aggregate to be worth it.

vadz commented:
I've force pushed my changes here now and will merge this as soon as the CI builds pass, if there are no objections.

Thanks again for your work on this!
